redfish/provider/data_source_redfish_firmware_inventory.go

```diff
@@ -1,173 +1,150 @@
+		Attributes:          FirmwareInventoryDatasourceSchema(),
+		Blocks:              RedfishServerDatasourceBlockMap(),
+	}
+}
+
+// FirmwareInventoryDatasourceSchema to define the Firmware Inventory data-source schema
+func FirmwareInventoryDatasourceSchema() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"id": schema.StringAttribute{
+			MarkdownDescription: "ID of the Firmware Inventory data-source",
+			Description:         "ID of the Firmware Inventory data-source",
+			Computed:            true,
+		},
+		"odata_id": schema.StringAttribute{
+			MarkdownDescription: "OData ID of the Firmware Inventory data-source",
+			Description:         "OData ID of the Firmware Inventory data-source",
+			Computed:            true,
+		},
+		"inventory": schema.ListNestedAttribute{
+			MarkdownDescription: "Firmware Inventory.",
+			Description:         "Firmware Inventory.",
+			Computed:            true,
+			NestedObject: schema.NestedAttributeObject{
+				Attributes: map[string]schema.Attribute{
+					"entity_name": schema.StringAttribute{
+						Computed:            true,
+						Description:         "entity name of the firmware inventory",
+						MarkdownDescription: "entity name of the firmware inventory",
+					},
+					"entity_id": schema.StringAttribute{
+						Computed:            true,
+						Description:         "entity id of the firmware inventory",
+						MarkdownDescription: "entity id of the firmware inventory",
+					},
+					"version": schema.StringAttribute{
+						Computed:            true,
+						Description:         "firmware inventory version",
+						MarkdownDescription: "firmware inventory version",
+					},
+				},
+			},
+		},
+	}
+}
+
+// Read implements datasource.DataSource
+func (g *FirmwareInventoryDatasource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var plan models.FirmwareInventory
+	diags := req.Config.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	service, err := NewConfig(g.p, &plan.RedfishServer)
+	if err != nil {
+		resp.Diagnostics.AddError("service error", err.Error())
+		return
+	}
+	state, err := readRedfishFirmwareInventory(service)
+	if err != nil {
+		diags.AddError("failed to fetch firmware inventory details", err.Error())
+	}
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func getInventoryItems(fwInventories []*redfish.SoftwareInventory) []models.Inventory {
+	var inv models.Inventory
+
+	inventoryItemList := make([]models.Inventory, 0)
+
+	for _, fwInv := range fwInventories {
+		if strings.HasPrefix(fwInv.Entity.ID, "Installed") {
+			inv.EntityId = types.StringValue(fwInv.Entity.ID)
+			inv.EntityName = types.StringValue(fwInv.Entity.Name)
+			inv.Version = types.StringValue(fwInv.Version)
+			inventoryItemList = append(inventoryItemList, inv)
+		}
```

Comment:
I would suggest de-indentation and restricting the scope of inv
```suggestion
		if !strings.HasPrefix(fwInv.Entity.ID, "Installed") {
			continue
		}
		inventoryItemList = append(inventoryItemList, models.Inventory{
			EntityId: types.StringValue(fwInv.Entity.ID)
			EntityName: types.StringValue(fwInv.Entity.Name)
			Version: types.StringValue(fwInv.Version)
		})
```